### PR TITLE
[Registration] add a step for organization selection in registration form

### PR DESCRIPTION
**Pull request by YanaMIvanova**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    | No

The following pull request adds a step for an organization choice in the registration form.

This makes it possible for a user to select the organization they belong to. This behaviour is controlled by a flag that is used to enable/disable the feature - `allow_organization_selection` (defaults to `false`)

If a chosen organization contains other organizations within it, another dropdown is shown for a further selection. 

On each selection, the organization in question is set in the Store as a `mainOrganization`, regardless of it being a parent or a child.

image

### CODE REVIEW

**Comment by maieutiquer** [comment]
Note: testing is awkward because of https://github.com/claroline/Claroline/issues/1727

File: src/main/core/Controller/APINew/User/RegistrationController.php [hunk 1]
```diff
@@ -165,6 +165,7 @@ public function initializeAction(Request $request): JsonResponse
                 'locale' => $request->getLocale(),
                 'forceOrganizationCreation' => $this->config->getParameter('registration.force_organization_creation'),
                 'allowWorkspace' => $this->config->getParameter('registration.allow_workspace'),
+                'allowOrganizationSelection' => $this->config->getParameter('registration.allow_organization_selection'),
```
Comment (Elorfin) [inline src/main/core/Controller/APINew/User/RegistrationController.php:168]: Can you expose the new option in the ui parameters please ? https://github.com/claroline/Claroline/blob/13.x/src/main/core/Resources/modules/administration/community/parameters/components/parameters-tab.jsx#L56

File: src/main/app/Resources/modules/security/registration/components/main.jsx [hunk 2]
```diff
@@ -61,6 +62,13 @@ class RegistrationMain extends Component {
       })
     }
 
+    if (this.props.options.allowOrganizationSelection) {
+      steps.push({
+        title: trans('organizations_selector'),
```
Comment (Elorfin) [inline src/main/app/Resources/modules/security/registration/components/main.jsx:67]: You can just reuse `trans('organization')`.

**Review by Elorfin — CHANGES_REQUESTED** [review]
IMHO, we should also disallow to use both options `forceOrganizationCreation` and `allowOrganizationSelection` at the same time. 

Both are used to fill the `mainOrganization` of the user and one will override the other.

File: src/main/app/Resources/modules/security/registration/components/organizationSelection.jsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+      const newSelectedOrganization = currentDepthOrganizations.find(org => org.code === selectedCode)
+
+      updateMainOrganization(newSelectedOrganization.name, selectedCode)
+    }
+
+    const selectedOrganization = currentDepthOrganizations.find(org => org.code === selectedCodes[depth])
+
+    return (
+      <>
```
[the remark this answers is not in the document]
Comment (YanaMIvanova) [reply src/main/app/Resources/modules/security/registration/components/organizationSelection.jsx]: Yes, it is the same. I changed it to `Fragment` for consistency

**Comment by YanaMIvanova** [comment]
> IMHO, we should also disallow to use both options `forceOrganizationCreation` and `allowOrganizationSelection` at the same time.
> 
> Both are used to fill the `mainOrganization` of the user and one will override the other.

Both options could be presented as radio buttons.
Would this be a proper behaviour? 

image

File: src/main/app/Resources/modules/security/registration/components/organizationSelection.jsx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            <>
+              {trans('selected_organizations_hierarchy')}
+              <ol>
+                {getOrganizationsNames(allOrganizations, selectedCodes).map(name => <li key={name}>{name}</li>)}
+              </ol>
+            </>
+          )}
+        </div>
+      </div>
+    </form>
```
[the remark this answers is not in the document]
Comment (YanaMIvanova) [reply src/main/app/Resources/modules/security/registration/components/organizationSelection.jsx]: `FormData` component seems to use the store for reading and writing (two-way data binding). As a result, every input in it shares the same content (the data in `mainOrganization.code`) 

image

[the remark this answers is not in the document]
Comment (Elorfin) [reply src/main/app/Resources/modules/security/registration/components/organizationSelection.jsx]: In fact you can use fields for props which does not belong to the form model (we usually add `_` as prefix for this). You can then add an `onChange(newValue)` on the input to update the final `mainOrganization` prop (which will be saved).

Here is an example of this kind of advanced configuration in forms : https://github.com/claroline/Claroline/blob/13.x/src/plugin/agenda/Resources/modules/events/event/components/form.jsx#L65.

**Comment by Elorfin** [comment]
> Both options could be presented as radio buttons.

This is a good idea. Both options should also be merged into one to simplify things. For example `organizationSelection: 'none'|'create'|'select'`.

We'll need an updater to set the new option for existing platforms. If you don't know how to do an Updater, let me know, I can do it for you.

File: src/main/app/Resources/modules/security/registration/components/organizationSelection.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (YanaMIvanova) [reply src/main/app/Resources/modules/security/registration/components/organizationSelection.jsx]: Thanks, I updated and pushed the form - now each input field has its own value

**Comment by YanaMIvanova** [comment]
I moved all organization options (`'none'|'create'|'select'`) into one prop

> We'll need an updater to set the new option for existing platforms. If you don't know how to do an Updater, let me know, I can do it for you.

Thanks, that would be really helpful

File: .gitignore [hunk 5]
```diff
@@ -32,3 +32,4 @@ package-lock.json
 /var/
 /vendor/
 ###< symfony/framework-bundle ###
+.idea
```
Comment (Elorfin) [inline .gitignore:35]: Should not have been commited. You should ignore it through your global `.gitignore`.

File: src/main/app/Resources/modules/security/registration/constants.js [hunk 6]
```diff
@@ -30,10 +30,24 @@ const registrationValidationTypes = {
   [REGISTRATION_MAIL_VALIDATION_FULL]: trans('force_mail_validation')
 }
 
+const ORGANIZATION_SELECTION_NONE = 'none'
+const ORGANIZATION_SELECTION_CREATE = 'create'
+const ORGANIZATION_SELECTION_SELECT = 'select'
+
+const ORGANIZATION_SELECTION_CHOICES = {
+  [ORGANIZATION_SELECTION_NONE]: 'No selection',
```
Comment (Elorfin) [inline src/main/app/Resources/modules/security/registration/constants.js:38]: Missing translation

**Review by Elorfin — APPROVED** [review]
👍

**Pull request merged**